scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,824 @@
+                    for (Iterator<Request> iterator = waitUntilQueue.iterator(); iterator.hasNext();) {
+                        Request waitRequest = iterator.next();
+                        if (testWaitFunction(service, waitRequest)) {
+                            service.serve(waitRequest);
+                            iterator.remove();
+                        }
+                    }
+                }
+
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    private <T> T evaluateClosure(String closureDefinition, Class<T> type) throws CompilationException {
+        try {
+            return (T) shell.evaluate(closureDefinition + " as " + type.getCanonicalName());
+        } catch (CompilationFailedException e) {
+            // CompilationFailedException contains instances which are not serializable
+            throw new CompilationException(StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    /**
+     * Execute the waitUntil function corresponding to the provided request
+     *
+     * If an error occurs, serve the request to unblock the caller with the exception attached
+     *
+     * @param service active object service
+     * @param request request to process
+     * @return true if the result of the waitUntil call is true or if an exception occurred
+     */
+    private boolean testWaitFunction(Service service, Request request) {
+        boolean dontNeedWaiting;
+        try {
+
+            dontNeedWaiting = executeWaitPredicateFunction(request);
+        } catch (Exception e) {
+            service.serveWithException(request, e);
+            // in case an exception occurs, don't delay the request, execute it immediately to notify the client
+            dontNeedWaiting = true;
+        }
+        return dontNeedWaiting;
+
+    }
+
+    /**
+     * Extract from the request the parameters corresponding to the waitUntil method call, then execute waitUntil with these parameters
+     * @param request request used to extract parameters
+     * @return the result of the waitUntil method call
+     * @throws InvocationTargetException
+     * @throws MethodCallExecutionFailedException
+     * @throws InvalidChannelException
+     * @throws CompilationException
+     */
+    private boolean executeWaitPredicateFunction(Request request) throws InvocationTargetException,
+            MethodCallExecutionFailedException, InvalidChannelException, CompilationException {
+        // extract the parameters which match the waitUntil call
+        int paramIndex = 0;
+        String originator = (String) request.getParameter(paramIndex++);
+        TaskId taskId = (TaskId) request.getParameter(paramIndex++);
+        String channel = (String) request.getParameter(paramIndex++);
+        String key = (String) request.getParameter(paramIndex++);
+        String predicate = (String) request.getParameter(paramIndex++);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable "paramIndex". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1854)